reconcile/utils/aws_api_typed/s3.py

```diff
@@ -0,0 +1,26 @@
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from mypy_boto3_s3 import S3Client
+    from mypy_boto3_s3.literals import BucketLocationConstraintType
+else:
+    S3Client = BucketLocationConstraintType = object
+
+
+class AWSApiS3:
+    def __init__(self, client: S3Client) -> None:
+        self.client = client
+
+    def create_bucket(self, name: str, region: str) -> str:
+        """Create an S3 bucket without any ACLs therefore the creator will be the owner and returns the ARN."""
+        bucket_kwargs = {}
+        if region != "us-east-1":
+            # you can't specify the location if it's us-east-1 :(
+            # see valid values "LocationConstraint" here: https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateBucketConfiguration.html
+            bucket_kwargs = {
+                "CreateBucketConfiguration": {
+                    "LocationConstraint": region,
+                },
+            }
+        self.client.create_bucket(Bucket=name, **bucket_kwargs)  # type: ignore
```

Comment:
so here we'll fail silently if the previous reconciliation loop created the bucket and sent the merge request, but it's not merged yet. Is that correct? Or do we raise an exception until the MR is merged?

What if the bucket creation fails for an other reason? (eg name used by any other AWS customer..)?

Reply:
Unfortunately or fortunately, boto3 `create_bucket` doesn't raise an exception if the bucket already exists, so it'll silently "fail" which is ok. If the bucket already exists, but for another AWS customer ((which is unlikely), we'll hit that error with the first usage of `terraform`. I think we can live with that.